File: mechanical/src/main/scala/resonantinduction/mechanical/process/grinder/TileGrindingWheel.java
package resonantinduction.mechanical.process.grinder;

import net.minecraft.block.material.Material;
import net.minecraft.entity.Entity;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.item.ItemBlock;
import net.minecraft.item.ItemStack;
import net.minecraft.util.DamageSource;
import net.minecraftforge.common.ForgeDirection;
import resonant.api.IRotatable;
import resonant.api.recipe.MachineRecipes;
import resonant.api.recipe.RecipeResource;
import resonant.lib.prefab.CustomDamageSource;
import resonant.lib.prefab.vector.Cuboid;
import resonantinduction.core.Reference;
import resonantinduction.core.ResonantInduction;
import resonantinduction.core.ResonantInduction.RecipeType;
import resonantinduction.core.interfaces.IMechanicalNode;
import resonantinduction.core.Timer;
import resonantinduction.mechanical.energy.grid.MechanicalNode;
import resonantinduction.mechanical.energy.grid.TileMechanical;
import universalelectricity.api.vector.Vector3;

/**
 * @author Calclavia
 */
public class TileGrindingWheel extends TileMechanical implements IRotatable
{
	public static final int PROCESS_TIME = 20 * 20;
	/**
	 * A map of ItemStacks and their remaining grind-time left.
	 */
	public static final Timer<EntityItem> timer = new Timer();

	public EntityItem grindingItem = null;

	private final long requiredTorque = 250;
	private double counter = 0;

	public TileGrindingWheel()
	{
		super(Material.rock);
		mechanicalNode = new GrinderNode(this).setLoad(2);
		bounds = new Cuboid(0.05f, 0.05f, 0.05f, 0.95f, 0.95f, 0.95f);
		isOpaqueCube = false;
		normalRender = false;
		customItemRender = true;
		rotationMask = Byte.parseByte("111111", 2);
		textureName = "material_steel_dark";
	}

	@Override
	public void updateEntity()
	{
		super.updateEntity();
		counter = Math.max(counter + Math.abs(mechanicalNode.torque), 0);
		doWork();
	}

	@Override
	public void collide(Entity entity)
	{
		if (entity instanceof EntityItem)
		{
			((EntityItem) entity).age--;
		}

		if (canWork())
		{
			if (entity instanceof EntityItem)
			{
				if (canGrind(((EntityItem) entity).getEntityItem()))
				{
					if (grindingItem == null)
					{
						grindingItem = (EntityItem) entity;
					}

					if (!TileGrindingWheel.timer.containsKey((EntityItem) entity))
					{
						TileGrindingWheel.timer.put((EntityItem) entity, TileGrindingWheel.PROCESS_TIME);
					}
				}
				else
				{
					entity.setPosition(entity.posX, entity.posY - 1.2, entity.posZ);
				}
			}
			else
			{
				entity.attackEntityFrom(new CustomDamageSource("grinder", this), 2);
			}

		}

		if (mechanicalNode.getAngularSpeed() != 0)
		{
			// Move entity based on the direction of the block.
			ForgeDirection dir = getDirection();
			dir = ForgeDirection.getOrientation(!(dir.ordinal() % 2 == 0) ? dir.ordinal() - 1 : dir.ordinal()).getOpposite();
			double speed = mechanicalNode.getAngularSpeed() / 20;
			double speedX = dir.offsetX * speed;
			double speedZ = dir.offsetZ * speed;
			double speedY = Math.random() * speed;
			if (Math.abs(speedX) > 1)
			{
				speedX = speedX > 0 ? 1 : -1;
			}
			if (Math.abs(speedZ) > 1)
			{
				speedZ = speedZ > 0 ? 1 : -1;
			}
			if (Math.abs(speedZ) > 1)
			{
				speedY = speedY > 0 ? 1 : -1;
			}
			entity.addVelocity(speedX, speedY, speedZ);
		}
	}

	/**
	 * Can this machine work this tick?
	 *
	 * @return
	 */
	public boolean canWork()
	{
		return counter >= requiredTorque;
	}

	public void doWork()
	{
		if (canWork())
		{
			boolean didWork = false;

			if (grindingItem != null)
			{
				if (timer.containsKey(grindingItem) && !grindingItem.isDead && new Vector3(this).add(0.5).distance(grindingItem) < 1)
				{
					int timeLeft = timer.decrease(grindingItem);

					if (timeLeft <= 0)
					{
						if (this.doGrind(grindingItem))
						{
							if (--grindingItem.getEntityItem().stackSize <= 0)
							{
								grindingItem.setDead();
								timer.remove(grindingItem);
								grindingItem = null;
							}
							else
							{
								grindingItem.setEntityItemStack(grindingItem.getEntityItem());
								// Reset timer
								timer.put(grindingItem, PROCESS_TIME);
							}
						}
					}
					else
					{
						grindingItem.delayBeforeCanPickup = 20;

						if (grindingItem.getEntityItem().getItem() instanceof ItemBlock)
						{
							ResonantInduction.proxy.renderBlockParticle(worldObj, new Vector3(grindingItem), new Vector3((Math.random() - 0.5f) * 3, (Math.random() - 0.5f) * 3, (Math.random() - 0.5f) * 3), ((ItemBlock) grindingItem.getEntityItem().getItem()).getBlockID(), 1);
						}
						else
						{
							worldObj.spawnParticle("crit", grindingItem.posX, grindingItem.posY, grindingItem.posZ, (Math.random() - 0.5f) * 3, (Math.random() - 0.5f) * 3, (Math.random() - 0.5f) * 3);
						}
					}

					didWork = true;
				}
				else
				{
					timer.remove(grindingItem);
					grindingItem = null;
				}
			}

			if (didWork)
			{
				if (this.ticks % 8 == 0)
				{
					worldObj.playSoundEffect(this.xCoord + 0.5, this.yCoord + 0.5, this.zCoord + 0.5, Reference.PREFIX + "grinder", 0.5f, 1);
				}

				counter -= requiredTorque;
			}
		}
	}

	public boolean canGrind(ItemStack itemStack)
	{
		return MachineRecipes.INSTANCE.getOutput(RecipeType.GRINDER.name(), itemStack).length > 0;
	}

	private boolean doGrind(EntityItem entity)
	{
		ItemStack itemStack = entity.getEntityItem();

		RecipeResource[] results = MachineRecipes.INSTANCE.getOutput(RecipeType.GRINDER.name(), itemStack);

		for (RecipeResource resource : results)
		{
			ItemStack outputStack = resource.getItemStack();

			if (!this.worldObj.isRemote)
			{
				EntityItem entityItem = new EntityItem(this.worldObj, entity.posX, entity.posY - 1.2, entity.posZ, outputStack);
				entityItem.delayBeforeCanPickup = 20;
				entityItem.motionX = 0;
				entityItem.motionY = 0;
				entityItem.motionZ = 0;
				this.worldObj.spawnEntityInWorld(entityItem);
			}
		}

		return results.length > 0;
	}

	@Override
	public ForgeDirection getDirection()
	{
		if (worldObj != null)
		{
			return ForgeDirection.getOrientation(getBlockMetadata());
		}

		return ForgeDirection.UNKNOWN;
	}

	@Override
	public void setDirection(ForgeDirection direction)
	{
		worldObj.setBlockMetadataWithNotify(xCoord, yCoord, zCoord, direction.ordinal(), 3);
	}

}

File: src/main/scala/resonantinduction/mechanical/mech/process/grinder/TileGrindingWheel.scala
package resonantinduction.mechanical.mech.process.grinder

import net.minecraft.block.material.Material
import net.minecraft.entity.Entity
import net.minecraft.entity.item.EntityItem
import net.minecraft.item.{ItemBlock, ItemStack}
import net.minecraftforge.common.util.ForgeDirection
import resonant.api.IRotatable
import resonant.api.recipe.{MachineRecipes, RecipeResource}
import resonant.content.factory.resources.RecipeType
import resonant.lib.prefab.damage.CustomDamageSource
import resonant.lib.transform.region.Cuboid
import resonant.lib.transform.vector.Vector3
import resonant.lib.utility.Timer
import resonantinduction.core.{Reference, ResonantInduction}
import resonantinduction.mechanical.mech.TileMechanical

/**
 * @author Calclavia
 */
object TileGrindingWheel
{
  final val processTime: Int = 20 * 20
  /**
   * A map of ItemStacks and their remaining grind-time left.
   */
  final val grindingTimer: Timer[EntityItem] = new Timer[EntityItem]
}

class TileGrindingWheel extends TileMechanical(Material.rock)
{
  private var grindingItem: EntityItem = null
  private final val requiredTorque: Long = 250
  private var counter = 0d

  mechanicalNode = new NodeGrinder(this)
  bounds = new Cuboid(0.05f, 0.05f, 0.05f, 0.95f, 0.95f, 0.95f)
  isOpaqueCube = false
  normalRender = false
  textureName = "material_steel_dark"

  override def update()
  {
    super.update()
    counter = Math.max(counter + Math.abs(mechanicalNode.torque), 0)
    doWork()

    if(!world.isRemote)
    {
      println(mechanicalNode)
      sendPacket(1)
    }
  }

  override def collide(entity: Entity)
  {
    if (entity.isInstanceOf[EntityItem])
    {
      entity.asInstanceOf[EntityItem].age -= 1
    }
    if (canWork)
    {
      if (entity.isInstanceOf[EntityItem])
      {
        if (canGrind((entity.asInstanceOf[EntityItem]).getEntityItem))
        {
          if (grindingItem == null)
          {
            grindingItem = entity.asInstanceOf[EntityItem]
          }
          if (!TileGrindingWheel.grindingTimer.containsKey(entity.asInstanceOf[EntityItem]))
          {
            TileGrindingWheel.grindingTimer.put(entity.asInstanceOf[EntityItem], TileGrindingWheel.processTime)
          }
        }
        else
        {
          entity.setPosition(entity.posX, entity.posY - 1.2, entity.posZ)
        }
      }
      else
      {
        entity.attackEntityFrom(new CustomDamageSource("grinder", this), 2)
      }
    }
    if (mechanicalNode.angularVelocity != 0)
    {
      var dir: ForgeDirection = getDirection
      dir = ForgeDirection.getOrientation(if (!(dir.ordinal % 2 == 0)) dir.ordinal - 1 else dir.ordinal).getOpposite
      val speed: Double = mechanicalNode.angularVelocity / 20
      var speedX: Double = dir.offsetX * speed
      var speedZ: Double = dir.offsetZ * speed
      var speedY: Double = Math.random * speed
      if (Math.abs(speedX) > 1)
      {
        speedX = if (speedX > 0) 1 else -1
      }
      if (Math.abs(speedZ) > 1)
      {
        speedZ = if (speedZ > 0) 1 else -1
      }
      if (Math.abs(speedZ) > 1)
      {
        speedY = if (speedY > 0) 1 else -1
      }
      entity.addVelocity(speedX, speedY, speedZ)
    }
  }

  /**
   * Can this machine work this tick?
   *
   * @return
   */
  def canWork: Boolean = counter >= requiredTorque

  def doWork()
  {
    if (canWork)
    {
      var didWork: Boolean = false
      if (grindingItem != null)
      {
        if (TileGrindingWheel.grindingTimer.containsKey(grindingItem) && !grindingItem.isDead && toVector3.add(0.5).distance(new Vector3(grindingItem)) < 1)
        {
          val timeLeft: Int = TileGrindingWheel.grindingTimer.decrease(grindingItem)
          if (timeLeft <= 0)
          {
            if (this.doGrind(grindingItem))
            {
              grindingItem.getEntityItem.stackSize -= 1;
              if (grindingItem.getEntityItem.stackSize <= 0)
              {
                grindingItem.setDead()
                TileGrindingWheel.grindingTimer.remove(grindingItem)
                grindingItem = null
              }
              else
              {
                grindingItem.setEntityItemStack(grindingItem.getEntityItem)
                TileGrindingWheel.grindingTimer.put(grindingItem, TileGrindingWheel.processTime)
              }
            }
          }
          else
          {
            grindingItem.delayBeforeCanPickup = 20
            if (grindingItem.getEntityItem.getItem.isInstanceOf[ItemBlock])
            {
              ResonantInduction.proxy.renderBlockParticle(worldObj, new Vector3(grindingItem), new Vector3((Math.random - 0.5f) * 3, (Math.random - 0.5f) * 3, (Math.random - 0.5f) * 3), 3, 1)
            }
            else
            {
              worldObj.spawnParticle("crit", grindingItem.posX, grindingItem.posY, grindingItem.posZ, (Math.random - 0.5f) * 3, (Math.random - 0.5f) * 3, (Math.random - 0.5f) * 3)
            }
          }
          didWork = true
        }
        else
        {
          TileGrindingWheel.grindingTimer.remove(grindingItem)
          grindingItem = null
        }
      }
      if (didWork)
      {
        if (this.ticks % 8 == 0)
        {
          worldObj.playSoundEffect(this.xCoord + 0.5, this.yCoord + 0.5, this.zCoord + 0.5, Reference.prefix + "grinder", 0.5f, 1)
        }
        counter -= requiredTorque
      }
    }
  }

  def canGrind(itemStack: ItemStack): Boolean =
  {
    return MachineRecipes.INSTANCE.getOutput(RecipeType.GRINDER.name, itemStack).length > 0
  }

  private def doGrind(entity: EntityItem): Boolean =
  {
    val itemStack: ItemStack = entity.getEntityItem
    val results: Array[RecipeResource] = MachineRecipes.INSTANCE.getOutput(RecipeType.GRINDER.name, itemStack)
    for (resource <- results)
    {
      val outputStack: ItemStack = resource.getItemStack
      if (!this.worldObj.isRemote)
      {
        val entityItem: EntityItem = new EntityItem(this.worldObj, entity.posX, entity.posY - 1.2, entity.posZ, outputStack)
        entityItem.delayBeforeCanPickup = 20
        entityItem.motionX = 0
        entityItem.motionY = 0
        entityItem.motionZ = 0
        this.worldObj.spawnEntityInWorld(entityItem)
      }
    }
    return results.length > 0
  }
}